libraries/AP_Math/scurves.cpp

```diff
@@ -0,0 +1,645 @@
+{
+    advance_time(time_scale * dt);
+    bool finish = move_from_time_pva_straight(_t, time_scale, pos, vel, accel);
+    pos -= _track;
+    return finish;
+}
+
+// return the position, velocity and acceleration vectors relative to the origin
+bool scurves::move_from_time_pva_straight(float time, float time_scale, Vector3f &pos, Vector3f &vel, Vector3f &accel)
+{
+    float scurve_P1, scurve_V1, scurve_A1, scurve_J1;
+    bool finish = update(time, scurve_J1, scurve_A1, scurve_V1, scurve_P1);
+    pos += _delta_unit_1 * scurve_P1;
+    vel += _delta_unit_1 * scurve_V1 * MIN(time_scale * 1.1f, 1.0f);
+    accel += _delta_unit_1 * scurve_A1 * time_scale;
+    return finish;
+}
+
+// Spline implementations
+
+// increment time pointer and return the position, velocity and acceleration vectors relative to the origin
+bool scurves::move_from_pva_spline(float dt, float time_scale, Vector3f &pos, Vector3f &vel, Vector3f &accel)
+{
+    advance_time(time_scale * dt);
+    bool finish = move_from_time_pva_spline(_t, time_scale, pos, vel, accel);
+    return finish;
+}
+
+// return the position, velocity and acceleration vectors relative to the destination
+bool scurves::move_to_pva_spline(float dt, float time_scale, Vector3f &pos, Vector3f &vel, Vector3f &accel)
+{
+    advance_time(time_scale * dt);
+    bool finish = move_from_time_pva_spline(_t, time_scale, pos, vel, accel);
+    pos -= _track;
+    return finish;
+}
+
+// return the position, velocity and acceleration vectors relative to the origin
+bool scurves::move_from_time_pva_spline(float time, float time_scale, Vector3f &pos, Vector3f &vel, Vector3f &accel)
+{
+    bool finish;
+    float scurve_P1, scurve_V1, scurve_A1, scurve_J1;
+    float time_start = segment[7].start_time;
+    float time_mid = segment[num_segs - 1].start_time - segment[11].start_time;
+
+    update(MIN(time, segment[11].start_time), scurve_J1, scurve_A1, scurve_V1, scurve_P1);
+    pos += _delta_unit_1 * scurve_P1;
+    vel += _delta_unit_1 * scurve_V1;
+    accel += _delta_unit_1 * scurve_A1;
+
+    update(MAX(time - time_start + segment[11].start_time, segment[11].start_time), scurve_J1, scurve_A1, scurve_V1, scurve_P1);
+    pos += _delta_unit_2 * (scurve_P1 - segment[11].start_pos);
+    vel += _delta_unit_2 * scurve_V1;
+    accel += _delta_unit_2 * scurve_A1;
+
+    update(MIN(segment[11].start_time - (time - time_start * 2 - time_mid + segment[11].start_time), segment[11].start_time), scurve_J1, scurve_A1, scurve_V1, scurve_P1);
+    pos += _delta_unit_3 * (segment[11].start_pos - scurve_P1);
+    vel += _delta_unit_3 * scurve_V1;
+    accel += _delta_unit_3 * (-scurve_A1);
+
+    finish = _t > (time_mid + 2.0 * time_start);
+    finish = _t > (time_mid + 2.0 * time_start);
+
+    vel *= MIN(time_scale * 1.1f, 1.0f);
+    accel *= time_scale;
+    return finish;
+}
+
+// debugging messages
+void scurves::debug()
```

Comment:
Should make sure this is being elided by the compiler.  Otherwise it needs to go behind an `#ifdef`

Reply:
debug hidden behind "#if 0" and moved closer to bottom of .h and .cpp files